### ISSUE TITLE
Opening a MessageBox to confirm a ListView selection change causes ListView to steal the event loop until the mouse is moved

### ISSUE THREAD

**Issue by daniels220** [dolphinsmalltalk/Dolphin#898]
Open something with a MethodBrowser—a class browser, method explorer, etc—with more than one method showing. Click on a method, make a change but don't accept it. Then carefully click on another method, without moving the mouse after you click. Answer "no" when asked if you want to save changes. The list selection will change, but the method workspace at the bottom doesn't update—until you move the mouse, even by a pixel, at which point it does.

I believe the process is getting stuck performing `aMouseEvent defaultWindowProcessing` in `ListView>>onButtonPressed:`—the ListView takes over the event loop waiting for a mouse-up event (while e.g. watching for movement to initiate a drag), but the mouse-up event has already been processed by the MessageBox. When the mouse is moved, I guess it notices that okay, I haven't seen a mouse up, but the mouse button *isn't down* so I better let go of the event loop. ...though this doesn't *exactly* fit, because if you move the mouse *while the MessageBox is still open*, the problem doesn't occur...

It's not obvious to me what the Right Way To Do This is, but opening a MessageBox to confirm a list selection change seems like something that other Win32 programmers would want to do, even outside of Dolphin, so I would *think* there must be a way to make it work...

**Comment by blairmcg** [comment]
You are right that this is something many Win32 programmers might want to do, but the ListView is not designed to support it in a consistent way. It doesn't have a native "selection changing" event - or at least not one that works consistently. [LVN_ITEMCHANGING](https://docs.microsoft.com/en-us/windows/win32/controls/lvn-itemchanging) notifications can be used to do this, but they are not sent when the control is in "virtual" ([LVS_OWNERDATA](https://docs.microsoft.com/en-us/windows/win32/controls/list-view-window-styles)) mode. We tend to favour the virtual mode as it is more flexible and generally offers better performance for larger lists. Therefore in order to provide a general "selection changing" event we deliberately went against an important principle we'd learned in trying to build previous frameworks using native widgets, which might be paraphrased as "Don't fight the control. You'll lose." In this case it worked out reasonably well, but your bug does rather validate the principle.
I'm not sure how much it matters though. Chances of encounter are low, and the impact when it happens also low. It may be fixable, but care would be needed not to break other scenarios or to rely too much on undocumented behaviour that may not work in future versions. So potentially costly/risky, and the value seems relatively low.

**Comment by blairmcg** [comment]
Actually, below is a fix for this. It needs a little more testing/refinement, but seems low risk.

```smalltalk
ListView>>onButtonPressed: aMouseEvent
	"Private - Common handler for a button down mouse event. Look for potential
	selection change events, and fire a #selectionChanging: event if necessary.
	The mouse event is only allowed to propagate to the control if we are happy that 
	the selection should indeed be changed."

	| indices |
	indices := self newSelectionsFromEvent: aMouseEvent.
	^indices = lastSelIndices
		ifTrue: [aMouseEvent defaultWindowProcessing]
		ifFalse: 
			[(self onSelChanging: indices cause: #mouse)
				ifTrue: 
					[| answer |
					"Selection change permitted?"
					(Keyboard default isButtonDown: aMouseEvent button)
						ifFalse: 
							["Post the mouse up consumed by a prompt that the control will miss"
							self
								postMessage: aMouseEvent message + 1
								wParam: 0
								lParam: aMouseEvent lParam].
					answer := aMouseEvent defaultWindowProcessing.
					self onSelChanged.	"Selection may have changed"
					answer]
				ifFalse: 
					["Suppress the mouse down so not received by control"
					0]]
```

**Comment by daniels220** [comment]
That makes sense about not fighting for control of the event loop. Unfortunate that such an obvious use-case wasn't considered in the design of the control, but I guess there's nothing we can do about it. (Well. One *could* theoretically reimplement everything that defaultWindowProcessing does, but designed to work within the normal, Dolphin-controlled event loop, but...no.)

Unfortunately...well, there's more to my story. Sounds like I'm probably going to have to find a specific workaround for my particular case, but here's the deal:

I'm actually encountering this most dramatically in automated tests, where there isn't any other interaction going on to unstick things and so the test hangs forever. The reason it's coming up now is that in Dolphin 6, merely having the mouse hovering over *any* Dolphin window (most conveniently the SmalltalkSystemShell) was enough to prevent things from sticking. And in older versions of Dolphin 7.0, I was able to work around the problem by posting a `WM_MOUSEMOVE` to the control—this was enough to unstick it iff the mouse button had in fact been released. Unfortunately, your proposed solution with a WM_[LMR]BUTTONUP message has an unwanted side-effect: it will (sometimes? always? often enough) cause the control to trigger an extra NM_CLICK notification, which ends up generating an extra `#selectionChanging:` event. This ends up being a problem for me because of arcane details of the testing environment, but duplicate notifications seem like a thing to avoid in any case.

So, I guess the only question I have at this point is, do you have any idea what might have changed, that previously the control would un-confuse itself (presumably inspecting the actual button state and realizing the button is no longer pressed?) on a WM_MOUSEMOVE, but now it only does so on a WM_[LMR]BUTTONUP?

**Comment by blairmcg** [comment]
As a quick response, try removing the line `self onSelChanged.	"Selection may have changed"`. That may now be duplicating the notification that the control is now sending.

**Comment by blairmcg** [comment]
A slightly firmer proposal would be something like 0a3edda8 in the branch blairmcg/898.
This ensures that the NM_CLICK/NM_RCLICK notifications are always generated and used to perform the actual selection change, and should mean there is only one selection change notification. It will need some time to bake even if it is the right thing to do.

To have any reasonable prospect of identifying changes that have caused the behaviour to change we'd need to narrow it down a bit, preferably by identifying the actual versions where the change occurred. Even using a binary-chop technique that can be time consuming. Given that this has always needed a workaround, I'm not sure that would be time well spent unless it really isn't possible to fix it.

**Comment by daniels220** [comment]
The duplicate event is a `#selectionChanging:` event, not a `#selectionChanged`—the one from `#onButtonPressed:`, and the one from `#maybeSelChanging:`. What "fixes" it for me—but this is a horrifically ugly hack—is to trick the control into thinking the selection-change has already happened while executing `#defaultWindowProcessing`, then undo the trickery before the final `#onSelChanged`:

```
onButtonPressed: aMouseEvent
	"Private - Common handler for a button down mouse event. Look for potential
	selection change events, and fire a #selectionChanging: event if necessary.
	The mouse event is only allowed to propagate to the control if we are happy that 
	the selection should indeed be changed."

	| indices |
	indices := self newSelectionsFromEvent: aMouseEvent.
	^indices = lastSelIndices
		ifTrue: [aMouseEvent defaultWindowProcessing]
		ifFalse: 
			[(self onSelChanging: indices cause: #mouse)
				ifTrue: 
					[| answer oldLastSelIndices |
					"Selection change permitted. Trick the control into thinking the selection has *already* changed,
					to prevent a duplicate #selectionChanging: event from #nmClick:"
					oldLastSelIndices := lastSelIndices.
					lastSelIndices := indices.
					answer := aMouseEvent defaultWindowProcessing.
					lastSelIndices := oldLastSelIndices.
					self onSelChanged.	"Selection may have changed"
					answer]
				ifFalse: 
					["Suppress the mouse down so not received by control"
					0]].
```

Perhaps this could be made less ugly by setting a flag for "selection change approved and in progress", and skipping additional events based on that rather than hacking `lastSelIndices`?

**Comment by blairmcg** [comment]
Ah, right. I'm assuming your code above would have the mouse up send in it?

A flag would be better, but like any transient mode there may be fragility. It would be an improvement to respond properly to the click notifications as the place that the selectionChanged: events are generated, but not selectionChanging:. The normal case is that selectionChanging: is fired only mouse down in order to allow the selection change to be blocked (this is where we are fighting what the list wants to do), but there is a special case of marquee selection in a multi-select list. That would have been firing a second selectionChanging: event, and I'm not sure that it should. The argument for doing so would be that the selections are different again. The argument against, that it is part of the same interaction and you certainly wouldn't want to prompt twice.  It needs some more thought/experimentation.

**Comment by daniels220** [comment]
Yes—when I saved that method I had the mouse-up send elsewhere, in my application code's #selectionChanging: handler, but it was there.

Yeah, I think you've pinned down the problem—it's the allowance for the possibility of a drag-select that's causing the duplicate event. My thoughts: The thing is that in that case, we don't really know the proper value for `#newSelections` until the drag finishes. We know that the selection is going to change—arguably "is changing"—but we don't know what it will be. Not sure how that should inform when we fire the event, but something to consider.

...Aha, something that might be helpful. It looks like that sort of drag-select can only be initiated by a click outside of any item, which we can easily detect in `#onButtonPressed:`. But I'm not sure exactly what to *do* with that information—to be useful we'd probably have to set a flag, and transient modes are fragile as you say... I'll sleep on it for tonight.

**Comment by blairmcg** [comment]
After not much thought it is apparent that it isn't possible to reconcile sending a single `selectionChanging:` event for marquee selection with the fact that this is sent on mouse down for mouse-based selections. Therefore I think we have to one of two things for marquee selection:
(i) Send two `selectionChanging:` events for marquee selection only; one for losing all selection when the mouse goes down outside any item to initiate the marquee selection, and one with the gained selections at the end of the drag. This is what happens currently in any case, and also it is likely one would only want to prompt to allow the selection change when the original selection is lost. That prompt would effectively cancel the marquee selection. When the 2nd event comes around, there will be no current selection.
(ii) Use the selection-reversion mechanism that is in-place to allow selection changes to be denied when made by keyboard actions, and for marquee selection delay the selection changing event until mouse up.
Although the 2nd option is then inconsistent with other mouse based selection in that the selection will visually change, and then revert, this seems preferable to raising two events for one user action, so I am leaning towards that. Marquee selection is a less important scenario anyway, particularly for the predominant `report mode` use of the list view.
I like the idea of processing the actual selection changes in response to the `nmClick:`/`nmRClick:` as is effectively done in my previous proposal that was always generating two `selectionChanging:` events. If that is kept, then it is just necessary to recognise that a `selectionChanging:` event has already been sent for the click and not send another, just the `selectionChanged:`. This would be similar to the way you suppressed it by temporarily modifying the `lastSelIndices`, but using separate state that holds the last confirmed new selections. This will avoid upsetting any of the existing uses of `lastSelIndices` (like the selection reversion mechanism), and so will fail in a safer way if something goes wrong or is wrong in the logic in some obscure case.
I merged some changes yesterday to allow for more inst vars in ListView.
The first thing to do, however, is to create some tests that replicate the existing problem, but also describe the expected sequence of events.

**Comment by blairmcg** [comment]
@daniels220, in your tests are you using a ListView in multi-select or single select mode? I ask because in the tests I am adding for the event sequence that simulate a click by posting events to the control I'm finding that in the case without a message box confirmation I get two selection changing events _anyway_ because of the `nmClick:` handler as it is currently implemented. What I'm seeing is the the control appears to grab control of the message loop when it is passed the `WM_LBUTTONDOWN` (as you said when reporting the issue), and then consumes and processes the `WM_LBUTTONUP` that my test has put in the queue and sends an `NM_CLICK` _before_ it returns from the `defaultWindowProcessing` call. Thus I get a 2nd `selectionChanging:` event before the `onButtonPressed:` method updates the selection.
It is very easy to demonstrate this by adding some diagnostic tracing at the top of the `ListView>>onSelChanging:cause:` method, like this:

```smalltalk
onSelChanging: anArray cause: aSymbol
	...
	Transcript print: self; nextPutAll: ' onSelChanging: '; print: anArray; nextPutAll: ' cause: '; print: aSymbol; cr.
	^self isStateRestoring or:
	...
```

With this in place, every time I click in the method list of a browser I see, for example:
```
a ListView(16r901AA, 'Methods') onSelChanging: #(218) cause: #mouse
a ListView(16r901AA, 'Methods') onSelChanging: #(218) cause: #mouse
```

The only case where the existing code does _not_ generate the 2nd `selectionChanging:` event for a multi-select list view is when a confirmation message box is shown. The effect of my previous proposal was to remove this exceptional case.

I'm intending to fix the duplicate `selectionChanging:` events (with the possible exception of marquee selection, where they won't be duplicates as such), but I'm curious as to why these were not already a problem in your tests.

**Comment by daniels220** [comment]
Yes, in my tests the view is in multi-select mode—but I realize that, although the test intercepts the MessageBox call and doesn't actually show one, another part of the code *does* process messages at that point (by way of a wait on a ModalMsgLoop). So, the mouse-up is consumed and `#nmClick:` doesn't fire. As I mentioned before, the reason I am encountering a new problem now is that, in Dolphin 7.0 from mid-2017, posting a `WM_MOUSEMOVE` was enough to unstick the control, but *without* firing `#nmClick:`, so that was what the test was doing. In turn in *production*, the `#nmClick:` does fire, but by that point the conditions that would cause the confirmation to be shown no longer apply (the mocking in the test is incomplete, which is why they continue to apply in the test—there are other tests that do simulate things more accurately, at the cost of performance, so it's a layered thing).

Thanks for working with me to wrangle this!

**Comment by blairmcg** [comment]
Ok, that is useful to know. The ListView documentation says that with respect to WM_LBUTTONDOWN it is: 

> Processed in different ways depending on whether a click or drag operation is being initiated. To determine which operation is involved, the list-view control enters a modal message loop until either the button is released or the mouse is moved. In the case of a click, the list-view control might change which item has the focus and which items are selected, taking into account the cursor position, the state of the SHIFT and CTRL keys, and so on. Then the list-view control sends its parent window an NM_CLICK (list view) notification code.
If dragging begins over an item, the list-view control selects and sets the focus to the item. Then it sends an LVN_BEGINDRAG notification code to the parent window. The parent window is responsible for actually carrying out the drag operation.
If dragging begins over the window background, the list-view control enters another modal message loop, enabling the user to form a rectangle by dragging the mouse. Items within the rectangle are selected.

So from that we would expect that either a button up or mouse move event should release the detection loop. I would expect there to be some hysteresis applied to the mouse move, and that it would then initiate a drag if moved by more than some amount with the mouse down. If so then a mouse move more than a certain distance would allow return from the WM_LBUTTONDOWN handler, a LVN_BEGINDRAG notification would be sent, and after that the ListView leaves it up to the app to implement the drag loop.

There is no immediately obvious reason that sending a WM_MOUSEMOVE would not cancel the mode. One would have to binary chop the history to find the commit where it goes wrong (if indeed it is due to a change in Dolphin and not Windows). As an experiment I tried posting a WM_MOUSEMOVE to the control in the `onButtonPressed:` in a similar way to the earlier modification to add a WM_MOUSEUP. I've pasted this method below. Running with this, when I try the manual scenario in a class browser that you outlined in your original posting, I find that the selection _does_ change after answering "No" the prompt. I don't think this is the right fix because I think it is better to receive and action the nmClick:, but it is interesting that the manual scenario is different than your test.

Actually I haven't yet managed to persuade the control to block in its WM_LBUTTONDOWN handler from my test, but when I do I'll then work on a fix.

```smalltalk
onButtonPressed: aMouseEvent
	"Private - Common handler for a button down mouse event. Look for potential selection change events, and fire a #selectionChanging: event if necessary. The mouse event is only allowed to propagate to the control if we are happy that the selection should indeed be changed."

	| indices |
	indices := self newSelectionsFromEvent: aMouseEvent.
	^indices = lastSelIndices
		ifTrue: [aMouseEvent defaultWindowProcessing]
		ifFalse: 
			[(self onSelChanging: indices cause: #mouse)
				ifTrue: 
					[| answer |
					"Selection change permitted"
					(Keyboard default isButtonDown: aMouseEvent button)
						ifFalse: 
							[self
								postMessage: WM_MOUSEMOVE
								wParam: (aMouseEvent wParam maskClear: aMouseEvent buttonFlag)
								lParam: aMouseEvent lParam].
					answer := aMouseEvent defaultWindowProcessing.
					self onSelChanged.	"Selection may have changed"
					answer]
				ifFalse: 
					["Suppress the mouse down so not received by control"
					0]]
```

**Comment by daniels220** [comment]
> But it only blocks if I move the mouse outside of any Dolphin window while the test runs, which is interesting

I have noticed this as well—in fact under older versions of Windows I was able to work around the issue by using `Cursor position:` to move the cursor over a Dolphin window before running such a test. This no longer seems to work, I think as a result of Windows being more protective of the user's control of the actual cursor position.

Anyway, the fix in #913 looks good—thanks for working on this!

**Comment by blairmcg** [comment]
That's good. 

> I have noticed this as well—in fact under older versions of Windows I was able to work around the issue by using Cursor position: to move the cursor over a Dolphin window before running such a test. This no longer seems to work, I think as a result of Windows being more protective of the user's control of the actual cursor position.

The tests actually do that in order to emulate the cursor being moved outside the window, and then move it back again to where it was (see `ListControlTest>>sendClickEvent:`). Windows should let the foreground app move the cursor, but not a background app. So if you've lost activation to something else (which probably includes Windows deciding the app is non-responsive) then it won't work.
